[scroll-animations] Plumb CSS originated animation range down to WebAnimation

nmoucht opened the pull request:
#### 70355b3509bd6ff1aac7944eff0dcfda27f9896b
<pre>
[scroll-animations] Plumb CSS originated animation range down to WebAnimation
<a href="https://bugs.webkit.org/show_bug.cgi?id=282059">https://bugs.webkit.org/show_bug.cgi?id=282059</a>
<a href="https://rdar.apple.com/138580596">rdar://138580596</a>

Reviewed by Antoine Quint.

Plumb CSS originated animation range down to WebAnimation. Add override bindings
setters to prevent CSS animation range from overwriting JS animation range.

* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::setRange):

Canonical link: <a href="https://commits.webkit.org/285707@main">https://commits.webkit.org/285707@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c4650a49c8ff016db32e6a58aff4a45287a34f55

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73528 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26336 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77805 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24766 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75643 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62088 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/742 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57781 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16207 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76595 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47879 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63265 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38191 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44492 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20751 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23097 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66276 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79414 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/327 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66172 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/985 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63277 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65451 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9311 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7487 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11336 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/807 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3563 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/837 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/823 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/843 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/CSSAnimation.cpp:
```diff
@@ -136,6 +136,8 @@ void CSSAnimation::syncPropertiesWithBackingAnimation()
         );
     }
 
+    setRange(animation.range());
```
You should either add a new overridden property type and follow it, or leave a note to had this (preferably with a matching bug) so we do it later. You can follow the pattern used elsewhere to add this overridden property which will requires adding a bindings-specific setter to `WebAnimation` to determine when the `rangeStart` and `rangeEnd` properties may have been set via the bindings.

nmoucht replied:
Made the setters binding specific and added check for if they have been set in `CSSAnimation::syncPropertiesWithBackingAnimation`.

graouts commented on Source/WebCore/animation/CSSAnimation.cpp:
```diff
@@ -134,6 +134,9 @@ void CSSAnimation::syncPropertiesWithBackingAnimation()
         );
     }
 
+    if (!m_overriddenProperties.contains(Property::Range))
+        setRange(animation.range());
```
This is good for now, but I think we may have to be able to differentiate between juste `rangeStart` or `rangeEnd` having been set via the JS API and let the other property be set via CSS. When we do, we should add test coverage for that if that isn't already the case.

nmoucht committed:
[scroll-animations] Plumb CSS originated animation range down to WebAnimation
https://bugs.webkit.org/show_bug.cgi?id=282059
rdar://138580596

Reviewed by Antoine Quint.

Plumb CSS originated animation range down to WebAnimation. Add override bindings
setters to prevent CSS animation range from overwriting JS animation range.

* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::setRange):

Canonical link: https://commits.webkit.org/285707@main
